solutions/usaco-696.mdx

```diff
@@ -2,13 +2,117 @@
 id: usaco-696
 source: USACO Platinum 2017 January
 title: Promotion Counting
-author: Benjamin Qi
+author: Benjamin Qi, Timothy Gao
 ---
 
-## Merging Indexed Sets
+[Official Editorial (Java)](http://www.usaco.org/current/data/sol_promote_platinum_jan17.html)
+
+## Solution 1: Euler Tour
+
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+Recall from the [Euler tour module](/gold/tree-euler) that each subtree is represented as a contiguous interval in its Euler tour. With this knowledge, we can compress the tree into an array. Now, the problem is reduced the following: for each node $i$, we need to determine the number of elements that are less than $\texttt{tour}[i]$ from $i$ to $\texttt{tout}[i]$, where $\texttt{tout}[i]$ is the end of the subtree interval for node $i$. By processing the nodes in decreasing order, this is a problem we can solve in $\mathcal{O}(N \log N)$ time with a binary-indexed tree.
+<LanguageSection>
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class PromotionCounting {
+	public static void main(String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new FileReader("promote.in"));
+		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("promote.out")));
+	 
+		int N = Integer.parseInt(br.readLine());
+		int[][] ratings = new int[N][2];
```

Comment:
might be better to make this a class for it to be more readable, but its probably fine since this is in plat already

Reply:
👍